# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [1]:
import pandas as pd
import numpy as np
from scipy import stats

# Challenge 1 - Independent Sample T-tests

In this challenge, we will be using the Pokemon dataset. Before applying statistical methods to this data, let's first examine the data.

To load the data, run the code below.

In [2]:
pokemon = pd.read_csv('../pokemon.csv')

Let's start off by looking at the `head` function in the cell below.

In [3]:
pokemon.head(3)

#       Name Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  \
0  1  Bulbasaur  Grass  Poison    318  45      49       49       65       65   
1  2    Ivysaur  Grass  Poison    405  60      62       63       80       80   
2  3   Venusaur  Grass  Poison    525  80      82       83      100      100   

   Speed  Generation  Legendary  
0     45           1      False  
1     60           1      False  
2     80           1      False

In [4]:
pokemon.Legendary.value_counts()

False    735
True      65
Name: Legendary, dtype: int64

The first thing we would like to do is compare the legendary Pokemon to the regular Pokemon. To do this, we should examine the data further. What is the count of legendary vs. non legendary Pokemons?

In [5]:
#  There are 735 Regular and 65 Legendary Pokemons

Compute the mean and standard deviation of the total points for both legendary and non-legendary Pokemon.

In [14]:
pokemon.groupby('Legendary')['Total'].agg([np.std, np.mean])

std        mean
Legendary                        
False      106.760417  417.213605
True        60.937389  637.384615

The computation of the mean might give us a clue regarding how the statistical test may turn out; However, it certainly does not prove whether there is a significant difference between the two groups.

In the cell below, use the `ttest_ind` function in `scipy.stats` to compare the the total points for legendary and non-legendary Pokemon. Since we do not have any information about the population, assume the variances are not equal.

In [10]:
legend=pokemon[pokemon['Legendary']==True]
non_legend=pokemon[pokemon['Legendary']==False]

What do you conclude from this test? Write your conclusions below.

In [15]:
stats.ttest_ind(legend['Total'], non_legend['Total'], equal_var=False)

Ttest_indResult(statistic=25.8335743895517, pvalue=9.357954335957446e-47)

In [ ]:
This test assumes that the populations DOES NOT have identical variances by default

H NULL: there is no diffreence between two groups.
H 1: there is a significant difference between average of two groups.
    
Degree of freedom for two group is nombre de observations N-1
Ttest_indResult(statistic=16.386116965872425, pvalue=3.0952457469652825e-52)

If the p-value is smaller than the threshold of 1% => we reject the null hypothesis of equal averages.

How about we try to compare the different types of pokemon? In the cell below, list the types of Pokemon from column `Type 1` and the count of each type.

In [16]:
pokemon['Type 1'].value_counts()

Water       112
Normal       98
Grass        70
Bug          69
Psychic      57
Fire         52
Electric     44
Rock         44
Ground       32
Dragon       32
Ghost        32
Dark         31
Poison       28
Fighting     27
Steel        27
Ice          24
Fairy        17
Flying        4
Name: Type 1, dtype: int64

Since water is the largest group of Pokemon, compare the mean and standard deviation of water Pokemon to all other Pokemon.

In [47]:
water=pokemon[pokemon['Type 1']=='Water']
other=pokemon[pokemon['Type 1'] != 'Water']

In [30]:
water_data=water['Total'].agg([np.mean, np.std])
water_data

mean    430.455357
std     113.188266
Name: Total, dtype: float64

In [31]:
other_data=other['Total'].agg([np.mean, np.std])
other_data

mean    435.859012
std     121.091682
Name: Total, dtype: float64

Perform a hypothesis test comparing the mean of total points for water Pokemon to all non-water Pokemon. Assume the variances are equal. 

In [44]:
stats.ttest_ind(water['Total'], other['Total'], equal_var=True)

Ttest_indResult(statistic=-0.4418547448849676, pvalue=0.6587140317488793)

Write your conclusion below.

In [ ]:
Ttest_indResult(statistic=-0.4418547448849676, pvalue=0.6587140317488793)

Hyp NULL :  the mean of total points for water Pokemon to all non-water Pokemon is same. 
Hyp 1 : the mean of total points for water Pokemon to all non-water Pokemon is NOT same. 
Assume the variances are equal.

pvalue=0.6587140317488793 => 65,9% which is > 10% threshold. It is in confidence interval
Hyp NULL is not to be rejected.

# 688 observations in OTHER group DF.
# 112 observations in WATER group DF

# Challenge 2 - Matched Pairs Test

In this challenge we will compare dependent samples of data describing our Pokemon. Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. In the cell below, import the `ttest_rel` function from `scipy.stats` and compare the two columns to see if there is a statistically significant difference between them.

In [52]:
stats.ttest_rel(pokemon['Defense'],pokemon['Attack'], axis=0)

Ttest_relResult(statistic=-4.325566393330478, pvalue=1.7140303479358558e-05)

Describe the results of the test in the cell below.

NULL hypothesis is that the defense and attack scores are equal for each pokemon.
In other words 2 related features (samples) of one pokemon have identical average (expected) values.

H 1 hypothesis is that the defense and attack scores are NOT equal for each pokemon.

Ttest_relResult(statistic=-4.325566393330478, pvalue=1.7140303479358558e-05)


pvalue is less close to 0%, so we do reject the Null Hypothesis.
Small p-values are associated with large t-statistics.

We are also curious about whether therer is a significant difference between the mean of special defense and the mean of special attack. Perform the hypothesis test in the cell below. 

In [54]:
stats.ttest_rel(pokemon['Sp. Atk'],pokemon['Sp. Def'], axis=0)

Ttest_relResult(statistic=0.853986188453353, pvalue=0.3933685997548122)

Describe the results of the test in the cell below.

NULL H: There is no difference between average of two features of pokemon.
H 1: There is difference between average of two features of pokemon.


Ttest_relResult(statistic=0.853986188453353, pvalue=0.3933685997548122)
pvalue is 39%, so we don't reject NULL HYP.
There is no difference between average of two features of pokemon.

As you may recall, a two sample matched pairs test can also be expressed as a one sample test of the difference between the two dependent columns.

Import the `ttest_1samp` function and perform a one sample t-test of the difference between defense and attack. Test the hypothesis that the difference between the means is zero. Confirm that the results of the test are the same.

In [57]:
pokemon['Defense'].mean()

73.8425

In [58]:
pokemon['Attack'].mean()

79.00125

In [64]:
stats.ttest_1samp(pokemon['Defense'], pokemon['Attack'].mean())

Ttest_1sampResult(statistic=-4.679124590910434, pvalue=3.3829080435248473e-06)

In [66]:
stats.ttest_1samp(pokemon['Attack'], pokemon['Defense'].mean())

Ttest_1sampResult(statistic=4.4954813919438585, pvalue=7.967950832362132e-06)

In [ ]:
NULL H: hypothesis that the difference between the means of two series is zero.
H 1:hypothesis that the difference between the means of two series is NOT zero.

"This is a two-sided test for the null hypothesis that the expected value
(mean) of a sample of independent observations a is equal to the given population mean, popmean.""

pvalue is in close to 0% and in the same range e-06 - > NULL H is rejected.

H 1: difference between the means of two series is NOT zero.

# Bonus Challenge - The Chi-Square Test

The Chi-Square test is used to determine whether there is a statistically significant difference in frequencies. In other words, we are testing whether there is a relationship between categorical variables or rather when the variables are independent. This test is an alternative to Fisher's exact test and is used in scenarios where sample sizes are larger. However, with a large enough sample size, both tests produce similar results. Read more about the Chi Squared test [here](https://en.wikipedia.org/wiki/Chi-squared_test).

In the cell below, create a contingency table using `pd.crosstab` comparing whether a Pokemon is legenadary or not and whether the Type 1 of a Pokemon is water or not.

In [ ]:
# Your code here:



Perform a chi-squared test using the `chi2_contingency` function in `scipy.stats`. You can read the documentation of the function [here](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.chi2_contingency.html).

In [ ]:
# Your code here:



Based on a 95% confidence, should we reject the null hypothesis?

In [ ]:
# Your answer here:

